### PR TITLE
Enable computing histograms using reduce intents

### PR DESCRIPTION
**Pull request by vasslitvinov**
A natural way to histogram something in parallel is something like this:

    const buckets = <a range or domain with an index per bucket>;
    var histo: [buckets] int;

    forall dataElement in <data to plot> with (+ reduce histo) {
      const bucket = <compute the bucket that dataElement falls into>;
      histo[bucket] += 1;
    }

    writeln(histo);

To make this work, we first equip chpl__sumType() in ChapelReduce.chpl
to handle arrays.

Then, a simple program like the first half of
   test/reductions/vass/histogram-1.chpl

would contain/execute more autoDestroys than autoCopies, corrupting memory.
The compiler changes in this PR fix that.

DETAILS

Turns out that when I generate an AST that MOVEs the result of an
array-returning function without autoCopy-ing it, then
returnRecordsByReferenceArguments() on this function does not fire
and things go off rail. Namely, the function itself and I think other
array-returning functions that it invokes get an extra autoDestroy
on the return value. This was the case for two 'svar's,
named shadowVar and shadowVarReduc, that stored the result of identity():

// before this change:

  redRef1->insertBefore("'move'(%S, identity(%S,%S))", // init
                        svar, gMethodToken, parentOp);

Adding autoCopy on the result resolves the issue.

As well as I replaced a PRIM_ASSIGN with a call to "=" on the result
of generate(). Since the reduction var already exists, we really want
assign into it, not bitcopy. To call "=", I had to put in "new CallExpr"
explicitly. Before I relied on insertAfter() to invoke new_Expr,
alas the latter does not handle calls to functions with operator names
like "=".

### CODE REVIEW

**Comment by vasslitvinov** [comment]
@diten or @mppf or @benharsh would you review?

File: modules/internal/ChapelReduce.chpl [hunk 1]
```diff
@@ -52,7 +52,16 @@ module ChapelReduce {
...
+    if isArray(x) {
+      type xET = x.eltType;
+      type xST = chpl__sumType(xET);
+      if xET == xST then
+        return eltType;
+      else
+        return [x.domain] xST;
+    } else {
+      return (x + x).type;
+    }
```
Comment (mppf) [inline modules/internal/ChapelReduce.chpl]: It concerns me that this code only applies to += reductions. Why? Shouldn't this work for a *= going through an array? Or anything else?
Why did you need to change this function at all? I would expect (x+x).type to work when x is an array.

File: compiler/resolution/implementForallIntents.cpp [hunk 2]
```diff
@@ -1374,12 +1378,16 @@ static void propagateExtraLeaderArgs(CallExpr* call, VarSymbol* retSym,
       ArgSymbol* parentOp = eFormal; // the reduceParent arg
       VarSymbol* currOp   = new VarSymbol(astrArg(ix, "reduceCurr"));
       VarSymbol* svar     = new VarSymbol(astrArg(ix, "shadowVar"));
+      VarSymbol* stemp    = newTemp("svTmp");
       redRef1->insertBefore(new DefExpr(currOp));
       redRef1->insertBefore("'move'(%S, clone(%S,%S))", // init
```
Comment (mppf) [inline compiler/resolution/implementForallIntents.cpp]: Can you explain why this one did not need an auto copy?

**Comment by mppf** [comment]
Hi @vasslitvinov - I asked some questions for specific lines. Other than that - I'd like you to check / test that in the example program (or in some related examples) the arrays/domains that are + reduced do not leak. Is the valgrind run clean for the example?

I don't feel particularly familiar with the code you are modifying.

What testing have you done?

File: modules/internal/ChapelReduce.chpl (hunk 1 above)
Comment (vasslitvinov) [inline modules/internal/ChapelReduce.chpl]: Good questions.

First, if (x+x).type worked, I wouldn't have bothered with this change. The reason it does not work is that (x+x) produces an _ir_ type, or anyway something similar. That does not work with the way the types of reduce-intent-related temps are inferred.

Why just for + ? That's because there is such a function only for +. The other operations are handled via _prod_id, _land_id, et al. (in modules/standard/Types.chpl). I suspect that those work for a smaller set of types than chpl__sumType().

File: compiler/resolution/implementForallIntents.cpp (hunk 2 above)
Comment (vasslitvinov) [inline compiler/resolution/implementForallIntents.cpp]: No, I can't. Things work the same way - i.e. valgrind reports no errors and leaks are the same amount - regardless of whether an autoCopy is present here or not. (I just checked.)

File: modules/internal/ChapelReduce.chpl (hunk 1 above)
Comment (vasslitvinov) [inline modules/internal/ChapelReduce.chpl]: P.S. I also verified that * reduction works on arrays.

**Comment by vasslitvinov** [comment]
Testing performed:
* linux64 and gasnet, wtih -futures
* valgrind on histogram-1.chpl and a variant with *=

Yes, memory is leaked on the arrays. Yet I consider these changes to be an improvement.

File: modules/internal/ChapelReduce.chpl (hunk 1 above)
Comment (mppf) [inline modules/internal/ChapelReduce.chpl]: Since we know that (x+x).type evaluates the LHS, what if it was:

  var tmp = x + x;
  return tmp.type;

**Comment by vasslitvinov** [comment]
As Michael points out, the test leaks the arrays. This is undesirable.
I will investigate the leaks. Closing this PR for now.

Other todos:
* follow up on Michael's other feedback
* add a test of *-reduction of arrays (which works and also leaks)
